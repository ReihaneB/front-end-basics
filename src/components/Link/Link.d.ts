import React from 'react';

interface LinkProps {
  to: string;
  children: React.ReactNode;
  autoFocus?: boolean;
  forwardedRef?: React.RefObject<HTMLButtonElement> | null;
}
