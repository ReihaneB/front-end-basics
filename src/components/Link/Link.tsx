import { memo } from 'react';
import NextLink from 'next/link';

import { useFocusable } from '@/hooks/useFocusable/useFocusable';

import type { LinkProps } from './Link.d';

function Link({
  to,
  children,
  autoFocus = false,
  forwardedRef = null,
  ...rest
}: LinkProps) {
  const { ref: elementRef } = useFocusable(forwardedRef, { autoFocus });

  return (
    <NextLink
      href={to}
      ref={elementRef}
      data-testid="link-component"
      {...rest}
    >
      {children}
    </NextLink>
  );
}

export default memo(Link);
