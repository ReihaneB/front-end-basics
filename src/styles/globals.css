:root {
  /* colors */
}

* {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
  border: 0;
  font: inherit;
  vertical-align: baseline;
}

html,
body {
  max-width: 100vw;
  overflow-x: hidden;
}

details,
figcaption,
figure,
footer,
hgroup,
menu,
nav {
  display: 'block',
}

ol,
ul {
  list-style: none;
}

blockquote,
q {
  quotes: none;
}

blockquote::before,
blockquote::after,
q::before,
q::after {
  content: '';
  content: none;
}

table {
  border-collapse: collapse;
  border-spacing: 0;
}

a {
  color: inherit;
  text-decoration: none;
}

h1 {
  font-size: 30px;
  line-height: 38px;
  font-weight: bold;
}

h2 {
  font-size: 20px;
  line-height: 24px;
  font-weight: bold;
}

h3 {
  font-size: 16px;
  line-height: 20px;
  font-weight: bold;
}

h4 {
  font-size: 14px;
  line-height: 18px;
}

h5 {
  font-size: 12px;
  line-height: 16px;
}

p {
  font-size: 16px;
  line-height: 24px;
}