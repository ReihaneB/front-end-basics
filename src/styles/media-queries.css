@custom-media --small-mobile (max-width: 375px);
@custom-media --mobile (max-width: 767px);
@custom-media --small-tablet (max-width: 957px);
@custom-media --tablet (max-width: 1199px);
@custom-media --ultra-wide (min-width: 1920px);