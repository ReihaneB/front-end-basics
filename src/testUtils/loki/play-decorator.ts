/* eslint-disable @typescript-eslint/ban-ts-comment */
// @ts-expect-error
import createAsyncCallback from '@loki/create-async-callback';
// @ts-expect-error
import isLokiRunning from '@loki/is-loki-running';
import type {PlayFunction} from '@storybook/csf';
// @ts-expect-error
import type {ReactFramework} from '@storybook/react';

async function waitForDocumentLoaded(): Promise<void> {
  if (document.readyState === 'loading') {
    return new Promise(resolve => {
      document.addEventListener('DOMContentLoaded', () => resolve());
    });
  }

  return Promise.resolve();
}

function enablePointerEvents(): void {
  const styleElement = document.createElement('style');
  document.documentElement.appendChild(styleElement);
  styleElement.sheet?.insertRule('* {pointer-events: auto !important}');
}

export function lokiPlayDecorator(
  target: PlayFunction<ReactFramework>
): PlayFunction<ReactFramework> {
  return async context => {
    const lokiRunning = isLokiRunning();
    const asyncCallback = createAsyncCallback();

    if (lokiRunning) {
      await waitForDocumentLoaded();
      enablePointerEvents();
    }

    try {
      await target(context);
    } finally {
      asyncCallback();
    }
  };
}
