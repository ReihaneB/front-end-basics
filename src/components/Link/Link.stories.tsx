import type { Meta, StoryObj } from '@storybook/react';
import { expect, userEvent, waitFor, within } from '@storybook/test';

import { lokiPlayDecorator } from '@/testUtils/loki/play-decorator';

import Link from './Link';

import styles from './Link.stories.module.css';

const meta: Meta<typeof Link> = {
  title: 'Components/Link',
  component: Link,
};

export default meta;
type Story = StoryObj<typeof Link>;

const defaultArgs = {
  to: '/',
  children: 'Click on the link',
  autoFocus: false,
  forwardedRef: null,
};

export const Playground: Story = {
  args: {
    ...defaultArgs,
  },
};

export const AutoFocus: Story = {
  args: {
    ...defaultArgs,
    autoFocus: true,
  },
  play: lokiPlayDecorator(
    async ({ canvasElement }) => {
      const canvas = within(canvasElement);

      await waitFor(() => expect(canvas.getByTestId('link-component')).toHaveFocus());
    }
  ),
};

export const CorrectHref: Story = {
  name: 'Correct href',
  args: {
    ...defaultArgs,
    children: '/link',
    to: '/link',
  },
  play: lokiPlayDecorator(
    async ({ args, canvasElement }) => {
      const canvas = within(canvasElement);

      expect(args.to).toBe('/link');
      expect(canvas.getByTestId('link-component')).toHaveAttribute('href', '/link');

      // TODO: Except Actions when it will be implemented with Storybook Interaction
    }
  ),
};

export const ClickMultipleTimes: Story = {
  args: {
    ...defaultArgs,
  },
  play: lokiPlayDecorator(
    async ({ canvasElement, step }) => {
      const canvas = within(canvasElement);

      await step('Click the link 2 times', async () => {
        await userEvent.click(canvas.getByTestId('link-component'));
        await userEvent.click(canvas.getByTestId('link-component'));
      });

      // TODO: Except Actions when it will be implemented with Storybook Interaction
    }
  ),
};

export const PressEnter: Story = {
  args: {
    ...defaultArgs,
  },
  play: lokiPlayDecorator(
    async ({ canvasElement, step }) => {
      const canvas = within(canvasElement);

      await step('Press the link 2 times', async () => {
        await userEvent.type(canvas.getByTestId('link-component'), '{Enter}');
        await userEvent.type(canvas.getByTestId('link-component'), '{Enter}');
      });

      // TODO: Except Actions when it will be implemented with Storybook Interaction
    }
  ),
};

export const FocusOnClick: Story = {
  name: 'Focused when clicked',
  args: {
    ...defaultArgs,
  },
  play: lokiPlayDecorator(
    async ({ canvasElement }) => {
      const canvas = within(canvasElement);

      await userEvent.click(canvas.getByTestId('link-component'));

      await waitFor(() => expect(canvas.getByTestId('link-component')).toHaveFocus());
    }
  ),
};

export const FocusOnTab: Story = {
  name: 'Focused when tabbed',
  args: {
    ...defaultArgs,
  },
  play: lokiPlayDecorator(
    async ({ canvasElement }) => {
      const canvas = within(canvasElement);

      await userEvent.tab();

      await waitFor(() => expect(canvas.getByTestId('link-component')).toHaveFocus());
    }
  ),
};

export function LongText() {
  return (
    <div className={styles.container}>
      <Link
        to="/"
      >
        This is a very long text that should be going to the next line
      </Link>
    </div>
  );
}

LongText.storyName = 'Long text';
