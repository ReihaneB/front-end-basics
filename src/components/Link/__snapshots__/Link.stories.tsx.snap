// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Link AutoFocus play-test 1`] = `
<main class="inter-normal">
  <a data-testid="link-component"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link ClickMultipleTimes play-test 1`] = `
<main class="inter-normal">
  <a data-testid="link-component"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link CorrectHref play-test 1`] = `
<main class="inter-normal">
  <a data-testid="link-component"
     href="/link"
  >
    /link
  </a>
</main>
`;

exports[`Components/Link FocusOnClick play-test 1`] = `
<main class="inter-normal">
  <a data-testid="link-component"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link FocusOnTab play-test 1`] = `
<main class="inter-normal">
  <a data-testid="link-component"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link LongText smoke-test 1`] = `
<main class="inter-normal">
  <div class="src-components-Link-Link-stories-module__container--Kpt8t">
    <a data-testid="link-component"
       href="/"
    >
      This is a very long text that should be going to the next line
    </a>
  </div>
</main>
`;

exports[`Components/Link Playground smoke-test 1`] = `
<main class="inter-normal">
  <a data-testid="link-component"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link PressEnter play-test 1`] = `
<main class="inter-normal">
  <a data-testid="link-component"
     href="/"
  >
    Click on the link
  </a>
</main>
`;
